### ISSUE TITLE
[PATCH] Module::CoreList provokes "uninitialized value" warning

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14403]
Migrated from [rt.perl.org#123556](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123556) (status was 'resolved')

Searchable as RT123556$

**Comment by p5pRT** [comment]
### From andreas@andreasvoegele.com

#### Created by andreas@andreasvoegele.com



If a script such as cpan2dist enables warnings with \-w\, warning
messages may be output when Module&#8203;::CoreList\->first\_release\(\) is
called with a module name and a version\.

 Use of uninitialized value in string ge at
 /usr/share/perl5/Module/CoreList\.pm line 30\.

The warning message is output if the module was originally added
to Perl without version information\, i\.e\. if its first version
in Module&#8203;::CoreList is "undef"\.  The following code snippet
reproduces the warning message\.

 use Module&#8203;::CoreList
 use English;
 local $WARNING = 1;
 print Module&#8203;::CoreList\->first\_release\('Getopt&#8203;::Long'\)\, "\\n";
 print Module&#8203;::CoreList\->first\_release\('Getopt&#8203;::Long'\, 2\.39\)\, "\\n";

No warning is output if the first call to "exists" in
first\_release\_raw\(\) is replaced by "defined" as shown in the
attached patch\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low
    Type=Patch
    PatchStatus=HasPatch
    module=Module::CoreList

Site configuration information for perl 5.21.8:

Configured by andreas at Tue Jan  6 12:11:38 CET 2015.

Summary of my perl5 (revision 5 version 21 subversion 8) configuration:
   
  Platform:
    osname=linux, osvers=3.14.27-vs2.3.6.13, archname=x86_64-linux
    uname='linux testing 3.14.27-vs2.3.6.13 #1 smp sat dec 20 12:09:12 cet 2014 x86_64 intel(r) core(tm)2 duo cpu t6670 @ 2.20ghz genuineintel gnulinux '
    config_args='-de -Dprefix=/usr -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl5 -Dsitearch=/usr/local/lib64/perl5 -Darchlib=/usr/lib64/perl5 -Dvendorprefix=/usr -Dprivlib=/usr/share/perl5 -Dvendorlib=/usr/share/perl5/vendor_perl -Dvendorarch=/usr/lib64/perl5/vendor_perl -Dscriptdir=/usr/bin -Dcccdlflags=-fPIC -Dinstallprefix=/usr -Dlibpth=/usr/local/lib64 /usr/lib64 /lib64 -Doptimize=-O2 -fPIC -D_FORTIFY_SOURCE=2 --param=ssp-buffer-size=4 -Wformat -Wformat-security -Werror=format-security -Duseshrplib -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr -Dusedevel -Darchname=x86_64-linux'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -fPIC -D_FORTIFY_SOURCE=2 --param=ssp-buffer-size=4 -Wformat -Wformat-security -Werror=format-security',
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.8.3', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib64 /usr/lib64 /lib64 /usr/local/lib /usr/lib64/gcc/x86_64-slackware-linux/4.8.3/include-fixed /usr/lib /lib/../lib64 /usr/lib/../lib64 /lib
    libs=-lpthread -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.20.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.20'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib64/perl5/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -fPIC -D_FORTIFY_SOURCE=2 --param=ssp-buffer-size=4 -Wformat -Wformat-security -Werror=format-security -L/usr/local/lib -fstack-protector'



@INC for perl 5.21.8:
    /usr/local/lib64/perl5
    /usr/local/share/perl5
    /usr/lib64/perl5/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib64/perl5
    /usr/share/perl5
    .


Environment for perl 5.21.8:
    HOME=/home/andreas
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=C
    LC_MESSAGES=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/bin:/bin:/usr/games:/opt/java/bin:/opt/java/jre/bin:/opt/scala/bin:/opt/sbt/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From andreas@andreasvoegele.com

<details><summary>0001-Only-compare-defined-values-with-ge-in.patch</summary>

```diff
From 336999c0d6e64ccbcdc2c2f6f64c47745a7f60b1 Mon Sep 17 00:00:00 2001
From: Andreas Voegele <andreas@andreasvoegele.com>
Date: Tue, 6 Jan 2015 12:05:14 +0100
Subject: [PATCH] Only compare defined values with "ge" in
 Module::CoreList::first_release_raw

---
 dist/Module-CoreList/lib/Module/CoreList.pm | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dist/Module-CoreList/lib/Module/CoreList.pm b/dist/Module-CoreList/lib/Module/CoreList.pm
index 654196b..6ade31e 100644
--- a/dist/Module-CoreList/lib/Module/CoreList.pm
+++ b/dist/Module-CoreList/lib/Module/CoreList.pm
@@ -28,7 +28,7 @@ sub first_release_raw {
     my $version = shift;
 
     my @perls = $version
-        ? grep { exists $version{$_}{ $module } &&
+        ? grep { defined $version{$_}{ $module } &&
                         $version{$_}{ $module } ge $version } keys %version
         : grep { exists $version{$_}{ $module }             } keys %version;
 
-- 
1.8.4


```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Jan 06 05&#8203;:33&#8203;:18 2015\, voegelas wrote&#8203;:
> This is a bug report for perl from andreas@&#8203;andreasvoegele\.com\,
> generated with the help of perlbug 1\.40 running under perl 5\.21\.8\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> If a script such as cpan2dist enables warnings with \-w\, warning
> messages may be output when Module&#8203;::CoreList\->first\_release\(\) is
> called with a module name and a version\.
> 
> Use of uninitialized value in string ge at
> /usr/share/perl5/Module/CoreList\.pm line 30\.
> 
> The warning message is output if the module was originally added
> to Perl without version information\, i\.e\. if its first version
> in Module&#8203;::CoreList is "undef"\.  The following code snippet
> reproduces the warning message\.
> 
> use Module&#8203;::CoreList
> use English;
> local $WARNING = 1;
> print Module&#8203;::CoreList\->first\_release\('Getopt&#8203;::Long'\)\, "\\n";
> print Module&#8203;::CoreList\->first\_release\('Getopt&#8203;::Long'\, 2\.39\)\, "\\n";
> 
> No warning is output if the first call to "exists" in
> first\_release\_raw\(\) is replaced by "defined" as shown in the
> attached patch\.
> 

Looks good\, but this is the sort of thing for which we should have a test in some file under dist/Module\-CoreList/t/

Thank you very much\.
Jim Keenan



\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From andreas@andreasvoegele.com

On Tue Jan 06 16&#8203;:00&#8203;:37 2015\, jkeenan wrote&#8203;:
> On Tue Jan 06 05&#8203;:33&#8203;:18 2015\, voegelas wrote&#8203;:
> > This is a bug report for perl from andreas@&#8203;andreasvoegele\.com\,
> > generated with the help of perlbug 1\.40 running under perl 5\.21\.8\.
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> >
> > If a script such as cpan2dist enables warnings with \-w\, warning
> > messages may be output when Module&#8203;::CoreList\->first\_release\(\) is
> > called with a module name and a version\.
> >
> > Use of uninitialized value in string ge at
> > /usr/share/perl5/Module/CoreList\.pm line 30\.
> > \[\.\.\.\]
> > No warning is output if the first call to "exists" in
> > first\_release\_raw\(\) is replaced by "defined" as shown in the
> > attached patch\.
> 
> Looks good\, but this is the sort of thing for which we should have a
> test in some file under dist/Module\-CoreList/t/

The attached patch adds a test to corelist\.t\. The test checks whether "strict" existed in Perl 5\.000 with an undefined version\. Then first\_release\('strict'\, 1\.01\) is called and the number of warnings is checked\.

Kind regards\,
Andreas

**Comment by p5pRT** [comment]
### From andreas@andreasvoegele.com

<details><summary>0001-Add-test-for-rt-123556.patch</summary>

```diff
From 726aca8fe4b7280dd9dff6734f0395a8777753ac Mon Sep 17 00:00:00 2001
From: Andreas Voegele <andreas@andreasvoegele.com>
Date: Wed, 7 Jan 2015 20:36:14 +0100
Subject: [PATCH] Add test for rt#123556

---
 dist/Module-CoreList/t/corelist.t | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/dist/Module-CoreList/t/corelist.t b/dist/Module-CoreList/t/corelist.t
index d8e93bf..c129584 100644
--- a/dist/Module-CoreList/t/corelist.t
+++ b/dist/Module-CoreList/t/corelist.t
@@ -1,7 +1,7 @@
 #!perl -w
 use strict;
 use Module::CoreList;
-use Test::More tests => 25;
+use Test::More tests => 29;
 
 BEGIN { require_ok('Module::CoreList'); }
 
@@ -89,3 +89,20 @@ is(Module::CoreList->removed_from('CPANPLUS::inc'), 5.010001,
 is(Module::CoreList::removed_from('CPANPLUS::inc'), 5.010001, 
    "CPANPLUS::inc was removed from 5.010001");
 
+{
+    my $warnings_count = 0;
+    local $SIG{__WARN__} = sub { $warnings_count++ };
+    local $^W = 1;
+
+    ok(exists $Module::CoreList::version{5}{strict},
+       "strict was in 5");
+
+    ok(!defined $Module::CoreList::version{5}{strict},
+       "strict had no version in 5");
+
+    is(Module::CoreList::first_release('strict', 1.01), 5.00405,
+       "strict reached 1.01 with 5.00405");
+
+    cmp_ok($warnings_count, '==', 0,
+           "an undefined version does not produce warnings rt#123556");
+}
-- 
1.8.4


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Jan 07 12&#8203;:02&#8203;:37 2015\, voegelas wrote&#8203;:
> On Tue Jan 06 16&#8203;:00&#8203;:37 2015\, jkeenan wrote&#8203;:
> > On Tue Jan 06 05&#8203;:33&#8203;:18 2015\, voegelas wrote&#8203;:
> > > This is a bug report for perl from andreas@&#8203;andreasvoegele\.com\,
> > > generated with the help of perlbug 1\.40 running under perl 5\.21\.8\.
> > >
> > > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > > \[Please describe your issue here\]
> > >
> > > If a script such as cpan2dist enables warnings with \-w\, warning
> > > messages may be output when Module&#8203;::CoreList\->first\_release\(\) is
> > > called with a module name and a version\.
> > >
> > > Use of uninitialized value in string ge at
> > > /usr/share/perl5/Module/CoreList\.pm line 30\.
> > > \[\.\.\.\]
> > > No warning is output if the first call to "exists" in
> > > first\_release\_raw\(\) is replaced by "defined" as shown in the
> > > attached patch\.
> >
> > Looks good\, but this is the sort of thing for which we should have a
> > test in some file under dist/Module\-CoreList/t/
> 
> The attached patch adds a test to corelist\.t\. The test checks whether
> "strict" existed in Perl 5\.000 with an undefined version\. Then
> first\_release\('strict'\, 1\.01\) is called and the number of warnings is
> checked\.

Thank you\.  Both patches applied\, as 41b9d3e and 17d1bb51f67\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
